// @shopify/prettier-plugin-liquid/dist/parser/grammar.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TAGS_WITHOUT_MARKUP = exports.VOID_ELEMENTS = exports.BLOCKS = exports.liquidStatementsGrammar = exports.liquidHtmlGrammar = exports.liquidHtmlGrammars = void 0;
const ohm_js_1 = __importDefault(require("ohm-js"));
exports.liquidHtmlGrammars = ohm_js_1.default.grammars(require('../../grammar/liquid-html.ohm.js'));
exports.liquidHtmlGrammar = exports.liquidHtmlGrammars['LiquidHTML'];
exports.liquidStatementsGrammar = exports.liquidHtmlGrammars['LiquidStatement'];
exports.BLOCKS = exports.liquidHtmlGrammar.rules.blockName.body.factors[0].terms.map((x) => x.obj);
exports.VOID_ELEMENTS = exports.liquidHtmlGrammar.rules.voidElementName.body.factors[0].terms.map((x) => x.args[0].obj);
exports.TAGS_WITHOUT_MARKUP = [
    'style',
    'schema',
    'javascript',
    'else',
    'break',
    'continue',
    'comment',
    'raw',
];
